Jint/Runtime/Debugger/DebugHandler.cs

```diff
@@ -1,85 +1,102 @@
+using System;
 using System.Collections.Generic;
 using System.Linq;
+using Esprima;
 using Esprima.Ast;
 using Jint.Native;
-using Jint.Native.Function;
-using Jint.Runtime.Descriptors;
 using Jint.Runtime.Environments;
-using Jint.Runtime.Interop;
 
 namespace Jint.Runtime.Debugger
 {
     internal class DebugHandler
     {
-        private readonly Stack<string> _debugCallStack;
+        private enum PauseType
+        {
+            Step,
+            Break
+        }
+
+        private bool _paused;
         private int _steppingDepth;
         private readonly Engine _engine;
 
         public DebugHandler(Engine engine)
         {
             _engine = engine;
-            _debugCallStack = new Stack<string>();
             _steppingDepth = int.MaxValue;
         }
 
-        internal void AddToDebugCallStack(JsValue function)
+        internal void OnStep(Statement statement)
         {
-            string name = GetCalleeName(function);
+            // Don't reenter if we're already paused (e.g. when evaluating a getter in a Break/Step handler)
+            if (_paused)
+            {
+                return;
+            }
 
-            _debugCallStack.Push(name);
-        }
+            Location location = statement.Location;
+            BreakPoint breakpoint = _engine.BreakPoints.FirstOrDefault(breakPoint => BpTest(location, breakPoint));
```

Comment:
not related to this PR but wonder why `Engine.BreakPoints` is a public list, probably would require better abstraction and offer `BpTest` as internal service instead of these context-capturing LINQ queries

Reply:
Must admit, I'm wondering why `BreakPoints` is even a property on `Engine`, which never uses it. 🙂 Other than for ease of access in API. But I might try to move it closer to the `DebugHandler` for this PR - now that it's breaking all other debugging APIs anyway.

Reply:
Sounds good 👍🏻

Reply:
Moved all debugger related properties, events etc. into `DebugHandler`, and turned that into a public property. The only "issue" with that is that the `sender` of Step/Break events should conventionally be `DebugHandler` then, but that's rather useless (but then, `sender` sometimes is...) - since it only has the two public events and the `BreakPoints` collection. What's the verdict on keeping the `sender` as `Engine`? Or should the two events and their invocations remain in `Engine`?